In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# https://www.kaggle.com/timmate/avocado-prices-2020

In [2]:
df = pd.read_csv('./data_csv/avocado-updated-2020.csv', parse_dates=['date'])

In [3]:
df.head()

date  average_price  total_volume       4046       4225      4770  \
0 2015-01-04           1.22      40873.28    2819.50   28287.42     49.90   
1 2015-01-04           1.79       1373.95      57.42     153.88      0.00   
2 2015-01-04           1.00     435021.49  364302.39   23821.16     82.15   
3 2015-01-04           1.76       3846.69    1500.15     938.35      0.00   
4 2015-01-04           1.08     788025.06   53987.31  552906.04  39995.03   

   total_bags  small_bags  large_bags  xlarge_bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     1162.65     1162.65        0.00          0.0       organic  2015   
2    46815.79    16707.15    30108.64          0.0  conventional  2015   
3     1408.19     1071.35      336.84          0.0       organic  2015   
4   141136.68   137146.07     3990.61          0.0  conventional  2015   

              geography  
0                Albany  
1                Albany  
2               Atlanta  
3               Atlanta  
4  Baltimore/Washington

In [4]:
2819.5+28287.42+49.9

31156.82

In [5]:
df.shape

(30021, 13)

In [6]:
df['year'].unique()

array([2015, 2016, 2017, 2018, 2019, 2020], dtype=int64)

In [7]:
jumlah = []

for i in df.columns:
    jumlah.append(df[i].nunique())

dfjumlah = pd.DataFrame(jumlah)
dfcol = pd.DataFrame(df.columns)  

unique = pd.concat([dfcol, dfjumlah], axis=1)
unique.columns = ['features', 'unique values']

unique

features  unique values
0            date            278
1   average_price            260
2    total_volume          29883
3            4046          28785
4            4225          29408
5            4770          19022
6      total_bags          29735
7      small_bags          28930
8      large_bags          25528
9     xlarge_bags          10295
10           type              2
11           year              6
12      geography             54

In [8]:
df[df['year'] == 2020].max()

date             2020-05-17 00:00:00
average_price                   2.34
total_volume             6.34289e+07
4046                     1.93089e+07
4225                     1.30364e+07
4770                     1.20586e+06
total_bags               3.16892e+07
small_bags               2.05504e+07
large_bags               1.33276e+07
xlarge_bags              1.02256e+06
type                         organic
year                            2020
geography        West Tex/New Mexico
dtype: object

In [9]:
df.isnull().sum()

date             0
average_price    0
total_volume     0
4046             0
4225             0
4770             0
total_bags       0
small_bags       0
large_bags       0
xlarge_bags      0
type             0
year             0
geography        0
dtype: int64

In [10]:
df.shape

(30021, 13)

In [11]:
df['type'].value_counts()

conventional    15012
organic         15009
Name: type, dtype: int64

In [12]:
df['average_price'].describe()

count    30021.000000
mean         1.390930
std          0.381347
min          0.440000
25%          1.110000
50%          1.350000
75%          1.630000
max          3.250000
Name: average_price, dtype: float64

In [13]:
df['year'].value_counts().sort_index()

2015    5615
2016    5616
2017    5722
2018    5292
2019    5616
2020    2160
Name: year, dtype: int64

In [14]:
df.head()

date  average_price  total_volume       4046       4225      4770  \
0 2015-01-04           1.22      40873.28    2819.50   28287.42     49.90   
1 2015-01-04           1.79       1373.95      57.42     153.88      0.00   
2 2015-01-04           1.00     435021.49  364302.39   23821.16     82.15   
3 2015-01-04           1.76       3846.69    1500.15     938.35      0.00   
4 2015-01-04           1.08     788025.06   53987.31  552906.04  39995.03   

   total_bags  small_bags  large_bags  xlarge_bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     1162.65     1162.65        0.00          0.0       organic  2015   
2    46815.79    16707.15    30108.64          0.0  conventional  2015   
3     1408.19     1071.35      336.84          0.0       organic  2015   
4   141136.68   137146.07     3990.61          0.0  conventional  2015   

              geography  
0                Albany  
1                Albany  
2               Atlanta  
3               Atlanta  
4  Baltimore/Washington

In [15]:
df.describe().T

count           mean           std      min       25%  \
average_price  30021.0       1.390930  3.813466e-01     0.44      1.11   
total_volume   30021.0  939254.592050  3.813519e+06    84.56  14298.53   
4046           30021.0  299107.331789  1.289108e+06     0.00    782.78   
4225           30021.0  284901.365483  1.169078e+06     0.00   2814.00   
4770           30021.0   21629.374703  1.009188e+05     0.00      0.00   
total_bags     30021.0  333533.629475  1.415618e+06     0.00   8373.82   
small_bags     30021.0  232125.625877  9.505028e+05     0.00   5955.98   
large_bags     30021.0   95185.285947  4.672104e+05     0.00    351.59   
xlarge_bags    30021.0    6222.717152  3.813737e+04     0.00      0.00   
year           30021.0    2017.205123  1.570133e+00  2015.00   2016.00   

                     50%        75%          max  
average_price       1.35       1.63         3.25  
total_volume   124204.74  489802.88  63716144.15  
4046            10522.76  115155.65  22743616.17  
4225            24567.22  140947.20  20470572.61  
4770              186.75    5424.16   2546439.11  
total_bags      50391.31  159174.17  31689188.60  
small_bags      34255.32  112938.44  20550406.59  
large_bags       5171.12   36067.62  13327600.76  
xlarge_bags         0.00     560.00   1022564.50  
year             2017.00    2019.00      2020.00

In [16]:
df.describe(include='O').T

count unique           top   freq
type       30021      2  conventional  15012
geography  30021     54    Pittsburgh    556

In [17]:
desc = []

for i in df.columns:
    desc.append([
        i,
        df[i].dtypes,
        df[i].isnull().sum(),
        round((((df[i].isnull().sum())/(len(df)))*100), 2),
        df[i].nunique(),
        df[i].drop_duplicates().sample(2).values
    ])

In [18]:
dfDesc = pd.DataFrame(desc, columns=['dataFeatures', 'dataType', 'null', 'nullPercentage', 'unique', 'uniqueSample'])

In [19]:
dfDesc

dataFeatures        dataType  null  nullPercentage  unique  \
0            date  datetime64[ns]     0             0.0     278   
1   average_price         float64     0             0.0     260   
2    total_volume         float64     0             0.0   29883   
3            4046         float64     0             0.0   28785   
4            4225         float64     0             0.0   29408   
5            4770         float64     0             0.0   19022   
6      total_bags         float64     0             0.0   29735   
7      small_bags         float64     0             0.0   28930   
8      large_bags         float64     0             0.0   25528   
9     xlarge_bags         float64     0             0.0   10295   
10           type          object     0             0.0       2   
11           year           int64     0             0.0       6   
12      geography          object     0             0.0      54   

                                         uniqueSample  
0   [2019-08-04T00:00:00.000000000, 2019-08-18T00:...  
1                                        [3.05, 2.12]  
2                                [26360.29, 41523.58]  
3                             [2956740.05, 148014.47]  
4                              [43424.52, 1354269.05]  
5                                 [107625.85, 4235.9]  
6                               [152983.63, 16819.08]  
7                                 [82964.19, 5548.04]  
8                                 [2394.52, 33276.69]  
9                                   [373.46, 8882.09]  
10                            [conventional, organic]  
11                                       [2017, 2020]  
12                                [Columbus, Seattle]

## insight data:

ada 13 kolom dalam dataset:
1. tanggal: dari januari 2015 - mei 2020
2. average_price: harga satuan avocado (US dollar)
3. total_volume: total penjualan avocado (lbs)
4. 4046
5. 4225
6. 4770 ==> total volume penjualan avocado PLU 4470 (perbedaan antara PLU 4046, 4225, 4770 adalah ukuran avocado)
7. total_bags: total kantong avocado terjual
8. small_bags: total penjualan avocado dengan pembelian menggunakan kantong kecil (tidak diketahui ukuran/dimensi kantong kecil)
9. large_bags: total penjualan avocado dengan pembelian menggunakan kantong large (tidak diketahui ukuran/dimensi kantong besar)
10. xlarge_bags: total penjualan avocado dengan pembelian menggunakan kantong extra large (tidak diketahui ukuran/dimensi kantong extra large)
11. type: tipe avocado (conventional dan organik)
12. year: tahun pencatatan transaksi: 2015 - 2020
13. geography: region kota tempat transaksi

- tidak ada data null dalam semua kolom
- dataset terdiri dari 13 kolom dan 30021 baris

Data dapat dilengkapi dengan keterangan Region di USA: https://www.nationalgeographic.org/maps/united-states-regions/#:~:text=A%20common%20way%20of%20referring,West%2C%20Southeast%2C%20and%20Midwest

## Problem statement:
Suatu online grocery shopping platform ingin mengekspansi wilayah layanannya (servis area) di kota di U.S. Berikan gambaran bagaimana kondisi pasar avocado di kota-kota U.S. Dapatkah algoritma machine learning membantu decision making perusahaan terhadap pertimbangan ekspansi ke kota baru?

## Tujuan:
- Analisis price elasticity dan volatility avocado di region-region, ataupun kota-kota di U.S
- Tentukan apakah pasar avocado di U.S dipengaruhi faktor musiman
- Buat machine learning model untuk memprediksi harga avocado guna membantu proses decision making perusahaan

## Data Preparation

terdapat beberapa kemungkinan kondisi pada dateset yang raw (mentah):
- adanya missing value
- adanya data yang tidak konsisten
- terdapat outliers
- data yang irrelevan terhadap target

sehingga pada tahap awal perlu dilakukan data cleaning, khususnya pengecekan terhadap adanya missing value dan data yang tidak konsisten.

missing value susah untuk diinterpretasikan oleh machine

### ** ekstrak data pada kolom 'date' **

In [20]:
# df['month'] = df['date'].dt.month

In [21]:
df['Date'] = df['date'].dt.day

In [22]:
df['month'] = df['date'].dt.month_name()

In [23]:
df.head()

date  average_price  total_volume       4046       4225      4770  \
0 2015-01-04           1.22      40873.28    2819.50   28287.42     49.90   
1 2015-01-04           1.79       1373.95      57.42     153.88      0.00   
2 2015-01-04           1.00     435021.49  364302.39   23821.16     82.15   
3 2015-01-04           1.76       3846.69    1500.15     938.35      0.00   
4 2015-01-04           1.08     788025.06   53987.31  552906.04  39995.03   

   total_bags  small_bags  large_bags  xlarge_bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1     1162.65     1162.65        0.00          0.0       organic  2015   
2    46815.79    16707.15    30108.64          0.0  conventional  2015   
3     1408.19     1071.35      336.84          0.0       organic  2015   
4   141136.68   137146.07     3990.61          0.0  conventional  2015   

              geography  Date    month  
0                Albany     4  January  
1                Albany     4  January  
2               Atlanta     4  January  
3               Atlanta     4  January  
4  Baltimore/Washington     4  January

In [24]:
df['month_num'] = df['date'].dt.month

In [25]:
df.drop(columns=['date'], inplace=True)

In [26]:
df.head()

average_price  total_volume       4046       4225      4770  total_bags  \
0           1.22      40873.28    2819.50   28287.42     49.90     9716.46   
1           1.79       1373.95      57.42     153.88      0.00     1162.65   
2           1.00     435021.49  364302.39   23821.16     82.15    46815.79   
3           1.76       3846.69    1500.15     938.35      0.00     1408.19   
4           1.08     788025.06   53987.31  552906.04  39995.03   141136.68   

   small_bags  large_bags  xlarge_bags          type  year  \
0     9186.93      529.53          0.0  conventional  2015   
1     1162.65        0.00          0.0       organic  2015   
2    16707.15    30108.64          0.0  conventional  2015   
3     1071.35      336.84          0.0       organic  2015   
4   137146.07     3990.61          0.0  conventional  2015   

              geography  Date    month  month_num  
0                Albany     4  January          1  
1                Albany     4  January          1  
2               Atlanta     4  January          1  
3               Atlanta     4  January          1  
4  Baltimore/Washington     4  January          1

### ** Tambah keterangan kolom region pada df **

In [27]:
print(df['geography'].unique())

['Albany' 'Atlanta' 'Baltimore/Washington' 'Boise' 'Boston'
 'Buffalo/Rochester' 'California' 'Charlotte' 'Chicago'
 'Cincinnati/Dayton' 'Columbus' 'Dallas/Ft. Worth' 'Denver' 'Detroit'
 'Grand Rapids' 'Great Lakes' 'Harrisburg/Scranton' 'Hartford/Springfield'
 'Houston' 'Indianapolis' 'Jacksonville' 'Las Vegas' 'Los Angeles'
 'Louisville' 'Miami/Ft. Lauderdale' 'Midsouth' 'Nashville'
 'New Orleans/Mobile' 'New York' 'Northeast' 'Northern New England'
 'Orlando' 'Philadelphia' 'Phoenix/Tucson' 'Pittsburgh' 'Plains'
 'Portland' 'Raleigh/Greensboro' 'Richmond/Norfolk' 'Roanoke' 'Sacramento'
 'San Diego' 'San Francisco' 'Seattle' 'South Carolina' 'South Central'
 'Southeast' 'Spokane' 'St. Louis' 'Syracuse' 'Tampa' 'Total U.S.' 'West'
 'West Tex/New Mexico']


In [28]:
df['geography'].nunique()

54

terdapat data yang inkonsisten dalam kolom 'geography' >> selain data nama-nama kota di AS, di kolom ini juga terdapat data 'Total_US', dan region-region di AS (Midsouth, Northeast, Southeast, South Central, West) >> didrop sehingga mendapatkan data keseluruhan hanya city

In [29]:
df_clean = df[(df['geography'] != 'Total U.S.') & (df['geography'] != 'Midsouth') & (df['geography'] != 'Northeast') & (df['geography'] != 'South Central') & (df['geography'] != 'Southeast')
& (df['geography'] != 'West')]

In [31]:
df_clean

average_price  total_volume       4046       4225      4770  \
0               1.22      40873.28    2819.50   28287.42     49.90   
1               1.79       1373.95      57.42     153.88      0.00   
2               1.00     435021.49  364302.39   23821.16     82.15   
3               1.76       3846.69    1500.15     938.35      0.00   
4               1.08     788025.06   53987.31  552906.04  39995.03   
...              ...           ...        ...        ...       ...   
30012           1.63       1798.00     249.00     143.00      0.00   
30013           1.12     591420.66  238039.66   16379.48    151.06   
30014           1.19      14134.00     200.00       0.00      0.00   
30019           0.89    1240709.05  430203.10  126497.28  21104.42   
30020           1.58      36881.00    1147.00    1243.00   2645.00   

       total_bags  small_bags  large_bags  xlarge_bags          type  year  \
0         9716.46     9186.93      529.53         0.00  conventional  2015   
1         1162.65     1162.65        0.00         0.00       organic  2015   
2        46815.79    16707.15    30108.64         0.00  conventional  2015   
3         1408.19     1071.35      336.84         0.00       organic  2015   
4       141136.68   137146.07     3990.61         0.00  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
30012     1407.00     1111.00      295.00         0.00       organic  2020   
30013   336846.03   276681.61    60134.42        30.00  conventional  2020   
30014    13933.00    13933.00        0.00         0.00       organic  2020   
30019   662904.25   395909.35   265177.09      1817.81  conventional  2020   
30020    31846.00    25621.00     6225.00         0.00       organic  2020   

                  geography  Date    month  month_num  
0                    Albany     4  January          1  
1                    Albany     4  January          1  
2                   Atlanta     4  January          1  
3                   Atlanta     4  January          1  
4      Baltimore/Washington     4  January          1  
...                     ...   ...      ...        ...  
30012              Syracuse    17      May          5  
30013                 Tampa    17      May          5  
30014                 Tampa    17      May          5  
30019   West Tex/New Mexico    17      May          5  
30020   West Tex/New Mexico    17      May          5  

[26685 rows x 15 columns]

In [32]:
df_clean['geography'].unique()

array(['Albany', 'Atlanta', 'Baltimore/Washington', 'Boise', 'Boston',
       'Buffalo/Rochester', 'California', 'Charlotte', 'Chicago',
       'Cincinnati/Dayton', 'Columbus', 'Dallas/Ft. Worth', 'Denver',
       'Detroit', 'Grand Rapids', 'Great Lakes', 'Harrisburg/Scranton',
       'Hartford/Springfield', 'Houston', 'Indianapolis', 'Jacksonville',
       'Las Vegas', 'Los Angeles', 'Louisville', 'Miami/Ft. Lauderdale',
       'Nashville', 'New Orleans/Mobile', 'New York',
       'Northern New England', 'Orlando', 'Philadelphia',
       'Phoenix/Tucson', 'Pittsburgh', 'Plains', 'Portland',
       'Raleigh/Greensboro', 'Richmond/Norfolk', 'Roanoke', 'Sacramento',
       'San Diego', 'San Francisco', 'Seattle', 'South Carolina',
       'Spokane', 'St. Louis', 'Syracuse', 'Tampa', 'West Tex/New Mexico'],
      dtype=object)

In [33]:
def region(df_clean):    
    if df_clean['geography'] == 'Albany' or df_clean['geography'] == 'Baltimore/Washington' or df_clean['geography'] == 'Boston' or df_clean['geography'] == 'Buffalo/Rochester' or df_clean['geography'] == 'Harrisburg/Scranton' or df_clean['geography'] == 'Hartford/Springfield' or df_clean['geography'] == 'New York' or df_clean['geography'] == 'Northern New England' or df_clean['geography'] == 'Philadelphia' or df_clean['geography'] == 'Pittsburgh' or df_clean['geography'] == 'Syracuse':
        return 'Northeastern'
    if df_clean['geography'] == 'Atlanta' or df_clean['geography'] == 'Charlotte' or df_clean['geography'] == 'Jacksonville' or df_clean['geography'] == 'Louisville' or df_clean['geography'] == 'Miami/Ft. Lauderdale' or df_clean['geography'] == 'Nashville' or df_clean['geography'] == 'New Orleans/Mobile' or df_clean['geography'] == 'Orlando' or df_clean['geography'] == 'Raleigh/Greensboro' or df_clean['geography'] == 'Richmond/Norfolk' or df_clean['geography'] == 'Roanoke' or df_clean['geography'] == 'South Carolina' or df_clean['geography'] == 'Tampa':
        return 'Southeastern'
    if df_clean['geography'] == 'Boise' or df_clean['geography'] == 'California' or df_clean['geography'] == 'Denver' or df_clean['geography'] == 'Las Vegas' or df_clean['geography'] == 'Los Angeles' or df_clean['geography'] == 'Plains' or df_clean['geography'] == 'Portland' or df_clean['geography'] == 'Sacramento' or df_clean['geography'] == 'San Diego' or df_clean['geography'] == 'San Francisco' or df_clean['geography'] == 'Seattle' or df_clean['geography'] == 'Spokane':
        return 'Western'
    if df_clean['geography'] == 'Chicago' or df_clean['geography'] == 'Cincinnati/Dayton' or df_clean['geography'] == 'Columbus' or df_clean['geography'] == 'Detroit' or df_clean['geography'] == 'Grand Rapids' or df_clean['geography'] == 'Great Lakes' or df_clean['geography'] == 'Indianapolis' or df_clean['geography'] == 'St. Louis':
        return 'Midwestern'
    if df_clean['geography'] == 'Dallas/Ft. Worth' or df_clean['geography'] == 'Houston' or df_clean['geography'] == 'Phoenix/Tucson' or df_clean['geography'] == 'West Tex/New Mexico':
        return 'Southwestern'
#     return 'Other'

In [34]:
# df_clean['region'] = df_clean.apply(lambda df_clean: region(df_clean), axis=1)

In [35]:
region = df_clean.apply(lambda df_clean: region(df_clean), axis=1)

In [36]:
df_full = pd.concat([df_clean, region], axis=1)

In [38]:
df_full.rename(columns={0: 'region', 'geography': 'city', 'Date': 'date'}, inplace=True)

In [39]:
df_full

average_price  total_volume       4046       4225      4770  \
0               1.22      40873.28    2819.50   28287.42     49.90   
1               1.79       1373.95      57.42     153.88      0.00   
2               1.00     435021.49  364302.39   23821.16     82.15   
3               1.76       3846.69    1500.15     938.35      0.00   
4               1.08     788025.06   53987.31  552906.04  39995.03   
...              ...           ...        ...        ...       ...   
30012           1.63       1798.00     249.00     143.00      0.00   
30013           1.12     591420.66  238039.66   16379.48    151.06   
30014           1.19      14134.00     200.00       0.00      0.00   
30019           0.89    1240709.05  430203.10  126497.28  21104.42   
30020           1.58      36881.00    1147.00    1243.00   2645.00   

       total_bags  small_bags  large_bags  xlarge_bags          type  year  \
0         9716.46     9186.93      529.53         0.00  conventional  2015   
1         1162.65     1162.65        0.00         0.00       organic  2015   
2        46815.79    16707.15    30108.64         0.00  conventional  2015   
3         1408.19     1071.35      336.84         0.00       organic  2015   
4       141136.68   137146.07     3990.61         0.00  conventional  2015   
...           ...         ...         ...          ...           ...   ...   
30012     1407.00     1111.00      295.00         0.00       organic  2020   
30013   336846.03   276681.61    60134.42        30.00  conventional  2020   
30014    13933.00    13933.00        0.00         0.00       organic  2020   
30019   662904.25   395909.35   265177.09      1817.81  conventional  2020   
30020    31846.00    25621.00     6225.00         0.00       organic  2020   

                       city  date    month  month_num        region  
0                    Albany     4  January          1  Northeastern  
1                    Albany     4  January          1  Northeastern  
2                   Atlanta     4  January          1  Southeastern  
3                   Atlanta     4  January          1  Southeastern  
4      Baltimore/Washington     4  January          1  Northeastern  
...                     ...   ...      ...        ...           ...  
30012              Syracuse    17      May          5  Northeastern  
30013                 Tampa    17      May          5  Southeastern  
30014                 Tampa    17      May          5  Southeastern  
30019   West Tex/New Mexico    17      May          5  Southwestern  
30020   West Tex/New Mexico    17      May          5  Southwestern  

[26685 rows x 16 columns]

In [40]:
# df_full.to_csv('./data_csv/updated_city_region.csv')